### ISSUE TITLE
[BUG] Cookie set by JavaScript it not sent after windows.location.reload(true)

### ISSUE THREAD

**Issue by pbiasotti** [microsoft/playwright#16945]
**Context:**
- Playwright Version: 1.24.1
- Operating System: Windows 11 and Ubuntu 20.04
- .NET version: NET 6
- Browser: Chromium
- Extra: The test fails when running in pipelines (not locally)

**Code Snippet**
N/A

**Describe the bug**

1. We have an MVC website which requires a cookie to render the razor view. 
2. For the first time a user gets to the site we detect that the user does not have the cookie in the MVC controller and set a cookie in the client side via javascript and reload the page

The razor view requires the timezone from the client side and the first time renders a script to set the cookie client side and reload the page
```
    if (timeZoneService.GetTimeZoneCookie().Equals(string.Empty))
    {
        executeTimeZone = true;
    }
    .....
    ....
    <script type="text/javascript">
        var shouldRun = "@executeTimeZone";

        if (shouldRun === 'True') {
            setTimeZone();
            window.location.reload(true);
        }
    </script> 
```

The `setTimezone` JS functions does something like the following to set the cookie:
```
     let cookieValue = "PreferredTimeZone=" + encodeURIComponent(element.value) + ";expires=" + encodeURIComponent(cookieExpiration) + "; path=/";
     document.cookie = cookieValue;
```

We have a simple test that hits the url of this page like:

```
await page.GotoAsync($"{baseUrl}");
```

The results are the following:

* **Locally**: Windows 11. Playwright 1.24.1. The test runs fine (headless or not) and when I see the traces in the traces viewer the reload request has the cookie `PreferredTimeZone` set by JS and the server detects this cookie and do not render the JS script again
* **Azure Pipelines**: Windows 11/Ubuntu-20.94. Playwright 1.24.1. When the test tries to browse the page it gets stuck with a lot of messages like the following and then times out:
```
waiting for navigation until "Load"
  navigated to "[url]
"Commit" event fired
  "Commit" event fired
  "Commit" event fired
  "Commit" event fired
  "Commit" event fired
  "Commit" event fired
  "Commit" event fired
  "Commit" event fired
  "Commit" event fired
  "Commit" event fired
  "Commit" event fired
  "Commit" event fired
  ....
  ============================================================
  ----> System.TimeoutException : Timeout 30000ms exceeded.
  Stack Trace:
     at Microsoft.Playwright.Core.Waiter.WaitForPromiseAsync[T](Task`1 task, Action dispose) in /_/src/Playwright/Core/Waiter.cs:line 179
   at Microsoft.Playwright.Core.Frame.WaitForNavigationAsync(FrameWaitForNavigationOptions options) in /_/src/Playwright/Core/Frame.cs:line 236
```

by looking at the traces the cookie set by Javascript is never send in the reload request which makes the razor view to re-render the scripts that reloads the page and this creates an infinite reload. 

**Note**: I'm new with playwright and I tried to make this issue as much clear as possible but probably I'm missing important information that could be good for troubleshooting. Anything I can test on my side or more information that I can provide I'll be happy to do it. For now we have the workaround of setting the cookie from the C# test itself but it would be nice to the get to the bottom of this to understand weather we should change our tests, or our site or there is a real bug somewhere in playwright.

**Comment by pavelfeldman** [comment]
When something works locally and does not work on CI, this often means that there is a race condition. How do you execute the snippet that sets the cookie? Do you do that from your test? Did you consider setting cookies in the storage instead using [context.AddCookiesAsync](https://playwright.dev/dotnet/docs/next/api/class-browsercontext#browser-context-add-cookies)?

**Comment by pbiasotti** [comment]
> When something works locally and does not work on CI, this often means that there is a race condition. How do you execute the snippet that sets the cookie? Do you do that from your test? Did you consider setting cookies in the storage instead using [context.AddCookiesAsync](https://playwright.dev/dotnet/docs/next/api/class-browsercontext#browser-context-add-cookies)?

In the real scenario the cookie is set by a JS script at the bottom of the HTML which simplified is something like the following :

```
    <script type="text/javascript">
       let cookieValue = "PreferredTimeZone=" + encodeURIComponent(element.value) + ";expires=" + encodeURIComponent(cookieExpiration) + "; path=/";
       document.cookie = cookieValue;
       window.location.reload(true);
    </script> 
```

To workaround the issue when running in the azure pipeline we are setting the cookie in our tests with the method [context.AddCookiesAsync](https://playwright.dev/dotnet/docs/next/api/class-browsercontext#browser-context-add-cookies) as you mentioned and it works, but this was just to unblock us to be able to add more tests and doesn't look like the real solution as we would expect the site to set the cookie itself.

Is there any debug logs that I could generate to check if this race condition exist and what is causing it?

**Comment by pavelfeldman** [comment]
> In the real scenario the cookie is set by a JS script at the bottom of the HTML which simplified is something like the following

Playwright isn't even involved here - you are setting the cookie inside the page and are reloading the page. So if your server does not receive the cookie with that reload request, it suggests that the browser itself is broken. That would be highly unlikely though. You should try isolating this and making sure there are no problems in the remainder of your test. No missing awaits, no concurrent requests, etc. If you could reproduce it locally and share the steps with us, that would help a lot. Otherwise we can't act on something that manifests itself as an improbable browser bug.

**Comment by dgozman** [comment]
Closing as per above. If you could provide a repro we can run locally, please file a new issue and link to this one.
